File: routes/sales.py
from flask import Blueprint, request, jsonify
from database import db, SalesOrder, SalesOrderItem, Product, InventoryLog, Customer
from datetime import datetime

sales_bp = Blueprint('sales', __name__)

def _order_dict(so):
    return {
        'id': so.id, 'order_no': so.order_no, 'customer_id': so.customer_id,
        'order_date': so.order_date, 'total_amount': so.total_amount,
        'paid_amount': so.paid_amount, 'status': so.status, 'notes': so.notes,
        'customer_name': so.customer.name
    }

@sales_bp.route('/sales-orders')
def list_sales_orders():
    search = request.args.get('search', '')
    status = request.args.get('status', '')
    q = SalesOrder.query.options(db.joinedload(SalesOrder.customer))
    if search:
        like = f'%{search}%'
        q = q.filter(db.or_(SalesOrder.order_no.like(like), Customer.name.like(like)))
    if status:
        q = q.filter(SalesOrder.status == status)
    orders = q.order_by(SalesOrder.created_at.desc()).all()
    return jsonify([_order_dict(so) for so in orders])

@sales_bp.route('/sales-orders/<int:order_id>')
def get_sales_order(order_id):
    so = SalesOrder.query.options(db.joinedload(SalesOrder.customer), db.joinedload(SalesOrder.items).joinedload(SalesOrderItem.product)).get(order_id)
    if not so:
        return jsonify({'error': '订单不存在'}), 404
    return jsonify({
        'order': _order_dict(so),
        'items': [{
            'id': i.id, 'sales_order_id': i.sales_order_id, 'product_id': i.product_id,
            'qty': i.qty, 'price': i.price, 'amount': i.amount,
            'product_name': i.product.name, 'product_code': i.product.code, 'unit': i.product.unit
        } for i in so.items]
    })

@sales_bp.route('/sales-orders', methods=['POST'])
def create_sales_order():
    data = request.json
    try:
        now = datetime.now()
        count = SalesOrder.query.count() + 1
        order_no = f'SO-{now.year}{now.month:02d}{now.day:02d}-{count:03d}'
        so = SalesOrder(
            order_no=order_no, customer_id=data['customer_id'],
            order_date=data['order_date'], total_amount=0,
            paid_amount=data.get('paid_amount', 0), status='draft',
            notes=data.get('notes', '')
        )
        db.session.add(so)
        db.session.flush()
        total = 0
        for item in data['items']:
            product = Product.query.get(item['product_id'])
            if not product:
                raise Exception(f"商品 {item['product_id']} 不存在")
            price = item['price']
            qty = int(item['qty'])
            amount = round(price * qty, 2)
            total += amount
            db.session.add(SalesOrderItem(sales_order_id=so.id, product_id=item['product_id'], qty=qty, price=price, amount=amount))
        so.total_amount = round(total, 2)
        db.session.commit()
        return jsonify({'id': so.id, 'order_no': order_no, 'message': '创建成功'}), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 400

@sales_bp.route('/sales-orders/<int:order_id>', methods=['PUT'])
def update_sales_order(order_id):
    data = request.json
    so = SalesOrder.query.options(db.joinedload(SalesOrder.items)).get(order_id)
    if not so:
        return jsonify({'error': '订单不存在'}), 404
    try:
        action = data.get('action', 'update')

        if action == 'confirm':
            if so.status != 'draft':
                raise Exception('只能确认草稿状态的订单')
            for item in so.items:
                product = Product.query.get(item.product_id)
                if product.stock_qty < item.qty:
                    raise Exception(f"商品 {product.name} 库存不足 (当前: {product.stock_qty}, 需要: {item.qty})")
                before_qty = product.stock_qty
                after_qty = before_qty - item.qty
                product.stock_qty = after_qty
                db.session.add(InventoryLog(product_id=item.product_id, change_type='sale', qty=item.qty, before_qty=before_qty, after_qty=after_qty, reference=so.order_no))
            so.status = 'confirmed'

        elif action == 'ship':
            if so.status != 'confirmed':
                raise Exception('只能发货已确认的订单')
            so.status = 'shipped'

        elif action == 'complete':
            if so.status not in ('confirmed', 'shipped'):
                raise Exception('只能完成已确认或已发货的订单')
            so.status = 'completed'
            so.paid_amount = so.total_amount

        elif action == 'cancel':
            if so.status == 'completed':
                raise Exception('已完成的订单不能取消')
            if so.status in ('confirmed', 'shipped'):
                for item in so.items:
                    product = Product.query.get(item.product_id)
                    before_qty = product.stock_qty
                    after_qty = before_qty + item.qty
                    product.stock_qty = after_qty
                    db.session.add(InventoryLog(product_id=item.product_id, change_type='cancel_return', qty=item.qty, before_qty=before_qty, after_qty=after_qty, reference=so.order_no))
            so.status = 'cancelled'

        elif action == 'update':
            if so.status != 'draft':
                raise Exception('只能编辑草稿状态的订单')
            so.customer_id = data.get('customer_id', so.customer_id)
            so.order_date = data.get('order_date', so.order_date)
            so.paid_amount = data.get('paid_amount', so.paid_amount)
            so.notes = data.get('notes', so.notes)
            if 'items' in data:
                SalesOrderItem.query.filter_by(sales_order_id=order_id).delete()
                total = 0
                for item in data['items']:
                    amount = round(item['price'] * item['qty'], 2)
                    total += amount
                    db.session.add(SalesOrderItem(sales_order_id=order_id, product_id=item['product_id'], qty=item['qty'], price=item['price'], amount=amount))
                so.total_amount = round(total, 2)

        db.session.commit()
        return jsonify({'message': '操作成功'})
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 400

@sales_bp.route('/sales-orders/<int:order_id>', methods=['DELETE'])
def delete_sales_order(order_id):
    so = SalesOrder.query.get(order_id)
    if not so:
        return jsonify({'error': '订单不存在'}), 404
    if so.status not in ('draft', 'cancelled'):
        return jsonify({'error': '只能删除草稿或已取消的订单'}), 400
    try:
        db.session.delete(so)
        db.session.commit()
        return jsonify({'message': '删除成功'})
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 400

File: static/js/sales.js
/* ========================================
   Sales Orders
   ======================================== */

const SALES_STATUS_FILTERS = [
    { value: '', label: '全部' },
    { value: 'draft', label: '草稿' },
    { value: 'confirmed', label: '已确认' },
    { value: 'shipped', label: '已发货' },
    { value: 'completed', label: '已完成' },
    { value: 'cancelled', label: '已取消' }
];

async function renderSalesList() {
    try {
        const orders = await api('/api/sales-orders');
        contentArea.innerHTML = `
            <div class="toolbar">
                <div class="toolbar-left">
                    <input type="text" class="search-input" id="sales-search" placeholder="🔍 搜索订单号/客户..." oninput="searchSales()">
                    <select class="form-control" id="sales-status-filter" onchange="searchSales()" style="width:130px">
                        ${SALES_STATUS_FILTERS.map(f => `<option value="${f.value}">${f.label}</option>`).join('')}
                    </select>
                </div>
                <div class="toolbar-right">
                    <button class="btn btn-primary" onclick="showSalesForm()">+ 新建销售订单</button>
                </div>
            </div>
            <div class="card">
                <div class="card-body no-padding">
                    <div class="table-wrap">
                        <table>
                            <thead><tr><th>订单号</th><th>客户</th><th>日期</th><th>金额</th><th>已付</th><th>状态</th><th>操作</th></tr></thead>
                            <tbody id="sales-tbody">${renderSalesOrderRows(orders)}</tbody>
                        </table>
                    </div>
                </div>
            </div>`;
    } catch (e) {
        contentArea.innerHTML = `<div class="alert alert-danger">加载失败: ${e.message}</div>`;
    }
}

function renderSalesOrderRows(orders) {
    if (!orders.length) return '<tr><td colspan="7" class="text-center text-muted">暂无数据</td></tr>';
    return orders.map(o => `
        <tr>
            <td class="font-bold">${o.order_no}</td>
            <td>${o.customer_name}</td>
            <td>${formatDate(o.order_date)}</td>
            <td>${formatMoney(o.total_amount)}</td>
            <td>${formatMoney(o.paid_amount)}</td>
            <td>${statusBadge(o.status)}</td>
            <td>
                ${o.status === 'draft' ? `<button class="btn btn-xs btn-success" onclick="changeSalesStatus(${o.id},'confirm')">确认</button>` : ''}
                ${o.status === 'confirmed' ? `<button class="btn btn-xs btn-warning" onclick="changeSalesStatus(${o.id},'ship')">发货</button>` : ''}
                ${o.status === 'confirmed' || o.status === 'shipped' ? `<button class="btn btn-xs btn-primary" onclick="changeSalesStatus(${o.id},'complete')">完成</button>` : ''}
                ${o.status !== 'completed' && o.status !== 'cancelled' ? `<button class="btn btn-xs btn-danger" onclick="changeSalesStatus(${o.id},'cancel')">取消</button>` : ''}
                ${o.status === 'draft' ? `<button class="btn btn-xs btn-secondary" onclick="showSalesForm(${o.id})">编辑</button>` : ''}
                ${o.status === 'draft' || o.status === 'cancelled' ? `<button class="btn btn-xs btn-danger" onclick="deleteSalesOrder(${o.id})">删除</button>` : ''}
            </td>
        </tr>`).join('');
}

async function searchSales() {
    const q = document.getElementById('sales-search').value;
    const status = document.getElementById('sales-status-filter').value;
    let url = '/api/sales-orders?';
    if (q) url += `search=${encodeURIComponent(q)}&`;
    if (status) url += `status=${status}&`;
    const orders = await api(url);
    document.getElementById('sales-tbody').innerHTML = renderSalesOrderRows(orders);
}

async function changeSalesStatus(id, action) {
    const labels = { confirm: '确认', ship: '发货', complete: '完成', cancel: '取消' };
    if (action === 'cancel' && !confirm('确定取消该订单吗？')) return;
    try {
        await api(`/api/sales-orders/${id}`, {
            method: 'PUT',
            body: JSON.stringify({ action })
        });
        showToast(`${labels[action]}成功`, 'success');
        renderSalesList();
    } catch (e) { /* handled */ }
}

async function deleteSalesOrder(id) {
    if (!confirm('确定删除该订单吗？')) return;
    try {
        await api(`/api/sales-orders/${id}`, { method: 'DELETE' });
        showToast('删除成功', 'success');
        renderSalesList();
    } catch (e) { /* handled */ }
}

async function showSalesForm(id) {
    contentArea.innerHTML = '<div class="empty-state"><div class="empty-icon">⏳</div><p>加载中...</p></div>';

    let order = null, items = [];
    if (id) {
        const data = await api(`/api/sales-orders/${id}`);
        order = data.order;
        items = data.items;
    }

    const [customers, products] = await Promise.all([
        api('/api/customers'),
        api('/api/products')
    ]);

    const isDraft = !order || order.status === 'draft';

    contentArea.innerHTML = `
        <div class="card mb-16">
            <div class="card-header flex-between">
                <h3>${id ? '编辑销售订单 - ' + order.order_no : '新建销售订单'}</h3>
                <button class="btn btn-secondary" onclick="renderSalesList()">← 返回列表</button>
            </div>
            <div class="card-body">
                <div class="form-row">
                    <div class="form-group">
                        <label>客户 <span class="required">*</span></label>
                        <select class="form-control" id="so-customer" ${!isDraft ? 'disabled' : ''}>
                            ${customers.map(c => `<option value="${c.id}" ${order && order.customer_id === c.id ? 'selected' : ''}>${c.name}</option>`).join('')}
                        </select>
                    </div>
                    <div class="form-group">
                        <label>订单日期 <span class="required">*</span></label>
                        <input class="form-control" type="date" id="so-date" value="${order ? order.order_date : new Date().toISOString().slice(0,10)}" ${!isDraft ? 'disabled' : ''}>
                    </div>
                </div>
                <div class="form-group">
                    <label>备注</label>
                    <input class="form-control" id="so-notes" value="${order ? order.notes || '' : ''}" ${!isDraft ? 'readonly' : ''}>
                </div>
            </div>
        </div>

        <div class="card mb-16">
            <div class="card-header flex-between">
                <h3>订单明细</h3>
                ${isDraft ? '<button class="btn btn-sm btn-primary" onclick="addSalesItem()">+ 添加商品</button>' : ''}
            </div>
            <div class="card-body no-padding">
                <div class="table-wrap">
                    <table class="order-items-table">
                        <thead><tr><th>商品</th><th>数量</th><th>单价</th><th>金额</th>${isDraft ? '<th>操作</th>' : ''}</tr></thead>
                        <tbody id="so-items">${renderSalesItems(items, products, isDraft)}</tbody>
                    </table>
                </div>
            </div>
            <div class="card-body">
                <div class="order-summary">合计: <span id="so-total">${formatMoney(order ? order.total_amount : 0)}</span></div>
            </div>
        </div>

        ${isDraft ? `
        <div style="text-align:right">
            <button class="btn btn-secondary" onclick="renderSalesList()">取消</button>
            <button class="btn btn-primary" onclick="saveSalesOrder(${id || 0})" style="margin-left:8px">
                ${id ? '更新订单' : '创建订单'}
            </button>
        </div>` : `
        <div class="alert alert-info">订单已${order.status === 'completed' ? '完成' : order.status === 'cancelled' ? '取消' : '确认'}，无法编辑</div>
        `}
    `;
}

function renderSalesItems(items, products, editable) {
    if (!items.length) return `<tr><td colspan="${editable ? 5 : 4}" class="text-center text-muted">暂无明细，请添加商品</td></tr>`;
    return items.map((item, i) => {
        const amount = (item.price || 0) * (item.qty || 0);
        return `
        <tr>
            <td>
                ${editable ? `<select class="form-control" onchange="updateSalesItemPrice(${i})" id="so-item-product-${i}">
                    ${products.map(p => `<option value="${p.id}" data-price="${p.sale_price}" ${item.product_id === p.id ? 'selected' : ''}>${p.name} (${p.code})</option>`).join('')}
                </select>` : (item.product_name || item.product_id)}
            </td>
            <td>${editable ? `<input type="number" value="${item.qty}" min="1" onchange="calcSalesTotal()" id="so-item-qty-${i}">` : item.qty}</td>
            <td>${editable ? `<input type="number" value="${item.price}" step="0.01" onchange="calcSalesTotal()" id="so-item-price-${i}">` : formatMoney(item.price)}</td>
            <td>${formatMoney(amount)}</td>
            ${editable ? `<td><button class="btn btn-xs btn-danger" onclick="removeSalesItem(${i})">删除</button></td>` : ''}
        </tr>`;
    }).join('');
}

function updateSalesItemPrice(i) {
    const sel = document.getElementById(`so-item-product-${i}`);
    const price = sel.options[sel.selectedIndex].dataset.price;
    document.getElementById(`so-item-price-${i}`).value = price;
    calcSalesTotal();
}

let salesItemCount = 0;

function addSalesItem() {
    const tbody = document.getElementById('so-items');
    salesItemCount = tbody.querySelectorAll('tr').length;
    const idx = salesItemCount;
    const row = document.createElement('tr');
    row.innerHTML = `
        <td><select class="form-control" id="so-item-product-${idx}" onchange="updateSalesItemPrice(${idx})"></select></td>
        <td><input type="number" value="1" min="1" onchange="calcSalesTotal()" id="so-item-qty-${idx}"></td>
        <td><input type="number" value="0" step="0.01" onchange="calcSalesTotal()" id="so-item-price-${idx}"></td>
        <td id="so-item-amount-${idx}">¥0.00</td>
        <td><button class="btn btn-xs btn-danger" onclick="this.closest('tr').remove();calcSalesTotal();">删除</button></td>`;
    tbody.appendChild(row);
    // Populate product select
    api('/api/products').then(products => {
        const sel = document.getElementById(`so-item-product-${idx}`);
        if (sel) {
            sel.innerHTML = products.map(p => `<option value="${p.id}" data-price="${p.sale_price}">${p.name} (${p.code})</option>`).join('');
            updateSalesItemPrice(idx);
        }
    });
}

function removeSalesItem(i) {
    const tr = document.querySelector(`#so-items tr:nth-child(${i + 1})`);
    if (tr) tr.remove();
    calcSalesTotal();
}

function calcSalesTotal() {
    const tbody = document.getElementById('so-items');
    if (!tbody) return;
    let total = 0;
    const rows = tbody.querySelectorAll('tr');
    rows.forEach((row, i) => {
        const qtyInput = row.querySelector(`[id^="so-item-qty-"]`);
        const priceInput = row.querySelector(`[id^="so-item-price-"]`);
        if (qtyInput && priceInput) {
            const qty = parseInt(qtyInput.value) || 0;
            const price = parseFloat(priceInput.value) || 0;
            total += qty * price;
        }
    });
    const totalEl = document.getElementById('so-total');
    if (totalEl) totalEl.textContent = formatMoney(total);
}

async function saveSalesOrder(id) {
    const customer_id = parseInt(document.getElementById('so-customer').value);
    const order_date = document.getElementById('so-date').value;
    const notes = document.getElementById('so-notes').value;

    if (!customer_id || !order_date) { showToast('请填写客户和日期', 'error'); return; }

    const items = [];
    const tbody = document.getElementById('so-items');
    const rows = tbody.querySelectorAll('tr');
    rows.forEach((row, i) => {
        const productSel = row.querySelector(`[id^="so-item-product-"]`);
        const qtyInput = row.querySelector(`[id^="so-item-qty-"]`);
        const priceInput = row.querySelector(`[id^="so-item-price-"]`);
        if (productSel && qtyInput && priceInput) {
            const product_id = parseInt(productSel.value);
            const qty = parseInt(qtyInput.value) || 0;
            const price = parseFloat(priceInput.value) || 0;
            if (product_id && qty > 0) items.push({ product_id, qty, price });
        }
    });

    if (!items.length) { showToast('请添加至少一个商品', 'error'); return; }

    try {
        if (id) {
            await api(`/api/sales-orders/${id}`, {
                method: 'PUT',
                body: JSON.stringify({ action: 'update', customer_id, order_date, notes, items })
            });
        } else {
            await api('/api/sales-orders', {
                method: 'POST',
                body: JSON.stringify({ customer_id, order_date, notes, items })
            });
        }
        showToast(id ? '更新成功' : '创建成功', 'success');
        renderSalesList();
    } catch (e) { /* handled */ }
}
